core/corehttp/gateway_handler.go

```diff
@@ -133,6 +133,7 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	w.Header().Set("X-IPFS-Path", urlPath)
+	w.Header().Set("Access-Control-Allow-Origin", "*")
```

Comment:
>> why would we likely have authentication on a gateway? wouldn't encryption of the resources themselves be the preferred method of keeping data safe on a p2p network?
>
> yes, but some people may want to enable simple auth. we shouldn't prevent them.

well of course we shouldn't prevent them, but it seems odd that we would want to encourage them to do that... unless there's some use-case that I'm missing. If people want to run it with an authentication layer in front, then they should probably reverse-proxy through a tool like nginx. Then they can choose whatever authentication layer they want without us cluttering up the ipfs core by adding options to do it for them.

> nginx and proxies are great, and certainly can be used this way. but we shouldn't force this to be the case.

Yeah, people shouldn't _normally_ need a proxy... go-ipfs should have reasonable defaults, such that you can use it for most things without needing to reverse-proxy. But if you're doing something uncommon like adding auth in front of it, then it seems reasonable to expect you to do a little setup.

> enabling CORS can allow lots of cross-site scripting problems.

such as?

> and users who just expose their gateway shouldnt be hit with it by not knowing their gateway could now be used to serve bad scripts to other sites. better safe here. users can choose to enable it and thus take on that responsibility.

But you can already do that without CORS? CORS is just there to allow browsers to use requests, initated from within a script or font-face declaration, and made avaliable to the script or renderer. We've always been able to load scripts from other domains without CORS... that's how JSONP works (which is basically a shim for CORS).

The only way you could prevent people from using a gateway to load malicious scripts is to whitelist specific files that are known to be safe... but that would kinda defeat the purpose of a P2P network... at that point you might as well just run a static file server with content-addressed URIs.

> We should endeavor to keep things consistent. there's all sorts of tools that need to be able to get the same headers. proxies, tunnels, headless browsers, pipes, ... i've definitely been annoyed in the past when a curl that i piped into an experimental browser-renderer didnt give me all the same headers it have chrome.

I'm not saying that we should send different headers to different tools. I'm saying that when cURL gets a header like `Access-Control-Allow-Origin: example.com`, it doesn't care, because it isn't a browser. CORS is only a restriction that is applied to browsers. When you write a program in Go, you are able to get content from whatever domain you want, regardless of where your Go program is stored, because only browsers care about CORS. However, when you write a program in JS and run it in a browser, you're only able to make requests to endpoints on the same domain as where your script was loaded from, or endpoints that have CORS enabled.

Thus, by sending a restrictive CORS header, you are only appling a restriction to browsers, not other programs. It does not apply to browsers and non-browsers alike. If there is a usecase for applying this kind of a restriction to only browsers, then it is a good idea to put something other than `Access-Control-Allow-Origin: *`. However, I think that is unreasonable to apply a restriction to browsers for a static resource like IPFS.